### DISCUSSION TITLE
Use recommended storage locations for app and user data

### DISCUSSION THREAD

**Discussion by danielniccoli** [microsoft/WSL#9191] [General]
WSL is putting application data directly in the user profile at `$env:userprofile\.wslconfig`. This violates Microsoft recommendations for Windows operating systems and should be moved to the recommended locations.

![image](https://user-images.githubusercontent.com/2971735/202553145-e6470fe4-973e-49c5-9e6c-097746c04c97.png)

The application and user data should be moved to storage locations recommended by Microsoft. A selection of recommended storage locations are

- ApplicationData (roaming)
- LocalApplicationData (non-roaming e.g., Local/LocalLow)
- CommonApplicationData (non-roaming)
- CommonProgramFiles (non-roaming)

Source and details about recommended storage locations:
https://docs.microsoft.com/en-us/dotnet/api/system.environment.specialfolder?view=net-6.0

Since `.wslconfig` may hold system-specific configurations, LocalApplicationData (`%LOCALAPPDATA%\wsl\config.ini`) seems to be the most fitting choice.

I would also suggest to rename `.wslconfig` to `config.ini` or something. The dot-naming is a unix-convention and not Windows-like.

**Comment by Biswa96** [comment]
> I would also suggest to rename .wslconfig to config.ini or something.

Renaming or moving the .wslconfig file would break many existing use cases.

**Comment by NotTheDr01ds** [comment]
I don't feel very strongly on this one myself, although I can understand the desire to get things of the Windows `$HOME` equivalent.  I *do* get frustrated with applications (even shells) that put their files in my Linux `$HOME` and much prefer the XDG `~/.config` standard.

It's just, with `$env:userprofile` on Windows, I don't use that directory much directly anyway.

> @Biswa96: Renaming or moving the `.wslconfig` file would break many existing use cases.

Not really.  Usually when an application changes the default location of a config file, they still maintain backwards compatibility by looking in the "old location" as a fallback.

An example would be `tmux`, which originally used `tmux.conf` (in the home directory), but now supports the XDG `~/.config/tmux/tmux.conf` (although unfortunately hardcoded to that directory last time I checked, rather than the proper `$XDG_CONFIG_HOME/tmux/tmux.conf`).

> @danielniccoli: I would also suggest to rename `.wslconfig` to `config.ini` or something.

Nooooo.  Not `config.ini` at least!  Too many DOS memories there 😉.

I'm actually okay with `.wslconfig` or (if it were in its own directory), `wslconfig`.  I'm typically editing it from inside WSL anyway, so I think of it as more a Linux config file than a Windows one.  I know a case may be made either way, but that's going to be more dependent on the individual's workflow.

> @danielniccoli: `%LOCALAPPDATA%\wsl\config.ini`

I don't mind that.  Just no deeper in the file structure please!  But what do I care? 😀-As long the `~/.wslconfig` is still read from its currently location as backup, I'll *probably* just keep using that one anyway.

**Comment by Biswa96** [comment]
I think the current `.wslconfig` file path is ok according to the rule. The link in OP says "Applications should not create files or folders at this level". The `.wslconfig` is not created by any WSL related programs but it is created by user, right? So, user can decide what to do with that file.

**Reply by pa-0** [reply]
Hate to resurrect a dead post, but I do believe this is a valid issue that should garner some attention / support.  

#### With respect to your response:

> but it is created by the user, right?

This would be perfectly acceptable, <ins>if</ins> the user actually had any choice in **_where_** they could store the file once they create it.  

As it currently stands currently, @danielniccoli's interpretation is the most sound: by offering this location **exclusively** as the valid save location for the configuration file, `wsl` effectively forces users to choose between:

(a) saving the config file in the location that Microsoft advises against, or 
(b) not setting any custom configurations for `wsl` at all.


I'm sure most would agree neither option seems very practical considering: 
- Microsoft's own recommendation, 
- a generally common preference held by many users and gaining more traction in popularity in recent years among users and developers alike to:
    - keep their $HOME directory clear from clutter
    - keep their configuration files in a single consolidated location for easy backup, portability, and versioning, and 
- the very basic principle that it is not unreasonable for a User to want _some level_ of discretion/control over the structure of, _at the very least_, one personal directory on a machine they own.

**Comment by danielniccoli** [comment]
> I think the current `.wslconfig` file path is ok according to the rule. The link in OP says "Applications should not create files or folders at this level". The `.wslconfig` is not created by any WSL related programs but it is created by user, right? So, user can decide what to do with that file.

You are taking this too literal. If application behaviour is controlled by a config file, it should live within a app-specific folder in `%LOCALAPPDATA%` or any of the other appdata folders. And a user cannot decide what to do with that file. If you need to customise WSL, it's either in `%USERPROFILE%\.wslconfig` or it's not doing anything. But that's irrelevant anyway, because an application should follow the basic guidelines set by an operating system and `%USERPROFILE%\.wslconfig` is anything but.

> Nooooo. Not config.ini at least! Too many DOS memories there 😉.

How a better name would look like would definitely be a topic of discussion. I will argue in favour of every name that follows Windows convention in the form of `<filename>.<extension>`.

**Comment by sryze** [comment]
Why can't Microsoft follow the decades old conventions for its own OS? I would like to not have any additional files in my profile folder please. And the dot naming convention does not really make sense as dot-files are not hidden automatically by Explorer or any other tools, unlike on Linux.

**Comment by danielniccoli** [comment]
> Why can't Microsoft follow the decades old conventions for its own OS?

Ignorance.
